File: client/src/api/cloud-export.ts
/**
 * API client for cloud export functionality
 */

import { apiClient } from '../lib/api';

// ─── Types ──────────────────────────────────────────────────

export interface CloudExportRequest {
  connection_id: number;
  source_path: string;
  cloud_folder?: string;
  link_type?: 'view' | 'edit';
  expires_at?: string | null;
}

export interface CloudExportJob {
  id: number;
  user_id: number;
  connection_id: number;
  source_path: string;
  file_name: string;
  is_directory: boolean;
  file_size_bytes: number | null;
  cloud_folder: string;
  cloud_path: string | null;
  share_link: string | null;
  link_type: string;
  status: 'pending' | 'uploading' | 'creating_link' | 'ready' | 'failed' | 'revoked';
  progress_bytes: number;
  error_message: string | null;
  created_at: string;
  completed_at: string | null;
  expires_at: string | null;
}

export interface CloudExportStatistics {
  total_exports: number;
  active_exports: number;
  failed_exports: number;
  total_upload_bytes: number;
}

export interface CheckScopeResponse {
  has_export_scope: boolean;
  provider: string;
}

// ─── API Functions ──────────────────────────────────────────

export async function startCloudExport(data: CloudExportRequest): Promise<CloudExportJob> {
  const resp = await apiClient.post('/api/cloud-export/', data);
  return resp.data;
}

export async function listCloudExports(limit = 50): Promise<CloudExportJob[]> {
  const resp = await apiClient.get('/api/cloud-export/jobs', { params: { limit } });
  return resp.data;
}

export async function getCloudExportStatus(jobId: number): Promise<CloudExportJob> {
  const resp = await apiClient.get(`/api/cloud-export/jobs/${jobId}`);
  return resp.data;
}

export async function revokeCloudExport(jobId: number): Promise<void> {
  await apiClient.post(`/api/cloud-export/jobs/${jobId}/revoke`);
}

export async function retryCloudExport(jobId: number): Promise<CloudExportJob> {
  const resp = await apiClient.post(`/api/cloud-export/jobs/${jobId}/retry`);
  return resp.data;
}

export async function getCloudExportStatistics(): Promise<CloudExportStatistics> {
  const resp = await apiClient.get('/api/cloud-export/statistics');
  return resp.data;
}

export async function checkConnectionScope(connectionId: number): Promise<CheckScopeResponse> {
  const resp = await apiClient.post('/api/cloud-export/check-scope', {
    connection_id: connectionId,
  });
  return resp.data;
}

File: backend/app/schemas/cloud_export.py
"""Pydantic schemas for cloud export API."""
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


class CloudExportRequest(BaseModel):
    """Request to start a cloud export job."""

    connection_id: int
    source_path: str = Field(min_length=1, description="Relative path on the NAS")
    cloud_folder: str = Field(
        default="BaluHost Shares/", description="Target folder in cloud drive"
    )
    link_type: Literal["view", "edit"] = Field(
        default="view", description="Sharing link permission type"
    )
    expires_at: Optional[datetime] = Field(
        default=None, description="Optional link expiration"
    )


class CloudExportJobResponse(BaseModel):
    """Response schema for an export job."""

    id: int
    user_id: int
    connection_id: int
    source_path: str
    file_name: str
    is_directory: bool
    file_size_bytes: Optional[int] = None
    cloud_folder: str
    cloud_path: Optional[str] = None
    share_link: Optional[str] = None
    link_type: str
    status: str
    progress_bytes: int
    error_message: Optional[str] = None
    created_at: datetime
    completed_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class CloudExportStatistics(BaseModel):
    """Statistics for the Cloud Shares tab on SharesPage."""

    total_exports: int
    active_exports: int
    failed_exports: int
    total_upload_bytes: int


class CheckScopeRequest(BaseModel):
    """Request to check if a connection has export-capable scope."""

    connection_id: int


class CheckScopeResponse(BaseModel):
    """Response indicating export scope availability."""

    has_export_scope: bool
    provider: str
